src/core/namespace.c

```diff
@@ -287,10 +288,15 @@ static int apply_mount(
 
                 /* First, get rid of everything that is below if there
                  * is anything... Then, overmount it with an
-                 * inaccessible directory. */
+                 * inaccessible path. */
                 umount_recursive(m->path, 0);
 
-                what = "/run/systemd/inaccessible";
+                r = lstat(m->path, &inaccessible_target);
```

Comment:
I think that both of you are right: path resolution follows symlinks, including intermediate ones, and chasing them across container chroot can be tricky (max number of symlinks, loops, multiple back-and-forth across chroot boundary). For this specific case, I'd propose to inspect path components below the chroot `root_directory` and error out if anything from there till the target is a symlink.

For the general case, I think there may be other cases of mismatches due to performing chroot after a user-controlled mount/stat/etc path-resolution (perhaps that should be separately explored to see if the order could be reversed, but I'm not familiar enough yet to foresee all the implications).